lib/puppet/functions/join.rb

```diff
@@ -0,0 +1,99 @@
+# Joins an array of values into a string with elements separated by a delimiter.
+#
+# Supports up to three arguments
+# * first argument is required and must be an `Array`, or an `Iterable`
+# * second arguments is the delimiter, must be a `String` if given, and defaults to an empty string.
+# * third argument is a Boolean, when if true will also format nested hashes
+#
+# @example Typical use of `join`
+#
+# ```puppet
+# join(['a','b','c'], ",")
+# # Would result in: "a,b,c"
+# ```
+#
+# Details:
+# * The array is flattened before elements are joined, but flattening does not extend to arrays nested in hashes or other objects unless
+#   the third argument is `true`
+# * When an `Iterable` is used (such as the result from `reverse_each`) it is converted to an array before joining. This
+#   means that it is also possible to join a hash as an array of nested tuples.
+# * In Puppet 5.5.0, the third argument was added to allow join of nested hashes as well as nested arrays.
```

Comment:
Using the third argument will also change the semantics from using the Ruby `#to_s` to `StringFormatter` with "%s". The documentation must mention that. This also raises the question of if it's ever desired to use Ruby #to_s in favor of the `StringFormatter` (the flag defaults to `false`). AFAIK, the former is not part of the Puppet Specifications.

If the intention is to provide a simpler API to the `StringFormatter` join functionality (which I would approve of), then perhaps pass a format as the third argument instead of a boolean? What if I want to join the result of formatting elements using '%p'?

Reply:
We must be 100% backwards compatible with the default - hence must use Ruby's `to_s` or we end up with unknown result (because we don't know what people return from their functions and we do not enforce that they are not Ruby runtime types, also because we would otherwise need to ensure string formatter produces 100% same output as Ruby for %s, which may not be wanted as that modifies the current use of string formatter).

I considered passing string formatter formats to join, but that snowballs as you want to control %s, %p, and you want to be able to specify spacing and delimiters. Instead I made the choice that the extended functionality arrayifies hashes. If you want more advanced formatting the String.new should instead be used. (My plan is to implement the other string formatting functions as calls to the string formatter.

Reply:
I agree about the backward compatibility. That's probably an issue that should be addressed separately and also affect functions like sprintf, notice, and others that rely on ruby-specific behavior.

The fact that the third parameter switches from ruby `#to_s` to `StringFormatter` semantics is still a problem IMO as it introduces an inconsistent behavior in the join function.